File: Gabriel_Code/Prototipo_0_Java/App/GraphHandler.java
package br.pucminas.grafo.handlers;

import br.pucminas.grafo.core.EventBus;
import br.pucminas.grafo.core.GraphRegistry;
import br.pucminas.grafo.events.Event;
import br.pucminas.grafo.events.EventType;
import br.pucminas.grafo.graph.AbstractGraph;
import br.pucminas.grafo.graph.AdjacencyListGraph;
import br.pucminas.grafo.graph.AdjacencyMatrixGraph;

/**
 * Handler responsável por toda operação estrutural sobre grafos.
 *
 * <p>Cada método é registrado como um handler no EventBus, respondendo
 * a um EventType específico — equivalente a um endpoint de API.</p>
 *
 * <h3>Eventos atendidos:</h3>
 * <ul>
 *   <li>GRAPH_CREATE        → payload: graphId, numVertices, impl ("matrix"|"list")</li>
 *   <li>GRAPH_ADD_EDGE      → payload: graphId, u, v</li>
 *   <li>GRAPH_REMOVE_EDGE   → payload: graphId, u, v</li>
 *   <li>GRAPH_HAS_EDGE      → payload: graphId, u, v → result: Boolean</li>
 *   <li>GRAPH_GET_VERTEX_COUNT → payload: graphId  → result: Integer</li>
 *   <li>GRAPH_GET_EDGE_COUNT   → payload: graphId  → result: Integer</li>
 *   <li>GRAPH_SET_VERTEX_WEIGHT → payload: graphId, v, weight</li>
 *   <li>GRAPH_GET_VERTEX_WEIGHT → payload: graphId, v → result: Double</li>
 *   <li>GRAPH_SET_EDGE_WEIGHT   → payload: graphId, u, v, weight</li>
 *   <li>GRAPH_GET_EDGE_WEIGHT   → payload: graphId, u, v → result: Double</li>
 *   <li>GRAPH_IN_DEGREE     → payload: graphId, v → result: Integer</li>
 *   <li>GRAPH_OUT_DEGREE    → payload: graphId, v → result: Integer</li>
 *   <li>GRAPH_IS_SUCCESSOR  → payload: graphId, u, v → result: Boolean</li>
 *   <li>GRAPH_IS_PREDECESSOR → payload: graphId, u, v → result: Boolean</li>
 *   <li>GRAPH_IS_DIVERGENT  → payload: graphId, u1,v1,u2,v2 → result: Boolean</li>
 *   <li>GRAPH_IS_CONVERGENT → payload: graphId, u1,v1,u2,v2 → result: Boolean</li>
 *   <li>GRAPH_IS_INCIDENT   → payload: graphId, u, v, x → result: Boolean</li>
 *   <li>GRAPH_IS_CONNECTED  → payload: graphId → result: Boolean</li>
 *   <li>GRAPH_IS_EMPTY      → payload: graphId → result: Boolean</li>
 *   <li>GRAPH_IS_COMPLETE   → payload: graphId → result: Boolean</li>
 *   <li>GRAPH_EXPORT_GEPHI  → payload: graphId, path</li>
 * </ul>
 */
public class GraphHandler {

    private final GraphRegistry registry;

    public GraphHandler(GraphRegistry registry) {
        this.registry = registry;
    }

    // ── Registro de handlers no bus ────────────────────────────────────────

    public void registerAll(EventBus bus) {
        bus.subscribe(EventType.GRAPH_CREATE,           this::onCreate);
        bus.subscribe(EventType.GRAPH_ADD_EDGE,         this::onAddEdge);
        bus.subscribe(EventType.GRAPH_REMOVE_EDGE,      this::onRemoveEdge);
        bus.subscribe(EventType.GRAPH_HAS_EDGE,         this::onHasEdge);
        bus.subscribe(EventType.GRAPH_GET_VERTEX_COUNT, this::onGetVertexCount);
        bus.subscribe(EventType.GRAPH_GET_EDGE_COUNT,   this::onGetEdgeCount);
        bus.subscribe(EventType.GRAPH_SET_VERTEX_WEIGHT,this::onSetVertexWeight);
        bus.subscribe(EventType.GRAPH_GET_VERTEX_WEIGHT,this::onGetVertexWeight);
        bus.subscribe(EventType.GRAPH_SET_EDGE_WEIGHT,  this::onSetEdgeWeight);
        bus.subscribe(EventType.GRAPH_GET_EDGE_WEIGHT,  this::onGetEdgeWeight);
        bus.subscribe(EventType.GRAPH_IN_DEGREE,        this::onInDegree);
        bus.subscribe(EventType.GRAPH_OUT_DEGREE,       this::onOutDegree);
        bus.subscribe(EventType.GRAPH_IS_SUCCESSOR,     this::onIsSuccessor);
        bus.subscribe(EventType.GRAPH_IS_PREDECESSOR,   this::onIsPredecessor);
        bus.subscribe(EventType.GRAPH_IS_DIVERGENT,     this::onIsDivergent);
        bus.subscribe(EventType.GRAPH_IS_CONVERGENT,    this::onIsConvergent);
        bus.subscribe(EventType.GRAPH_IS_INCIDENT,      this::onIsIncident);
        bus.subscribe(EventType.GRAPH_IS_CONNECTED,     this::onIsConnected);
        bus.subscribe(EventType.GRAPH_IS_EMPTY,         this::onIsEmpty);
        bus.subscribe(EventType.GRAPH_IS_COMPLETE,      this::onIsComplete);
        bus.subscribe(EventType.GRAPH_ADD_VERTEX,       this::onAddVertex);
        bus.subscribe(EventType.GRAPH_EXPORT_GEPHI,     this::onExportGephi);
    }

    // ── Handlers ───────────────────────────────────────────────────────────

    private void onCreate(Event ev) {
        String id   = ev.getString("graphId");
        int    n    = ev.getInt("numVertices");
        String impl = ev.getString("impl");

        if (registry.contains(id)) {
            ev.setError("Grafo já existe: " + id);
            return;
        }
        AbstractGraph g = "matrix".equalsIgnoreCase(impl)
                ? new AdjacencyMatrixGraph(n)
                : new AdjacencyListGraph(n);
        registry.register(id, g);
        ev.setResult(g);
    }

    /** Evento GRAPH_ADD_VERTEX: expande o grafo existente criando um novo com +1 vértice.
     *  Payload: graphId, label (opcional). Result: novo índice do vértice. */
    private void onAddVertex(Event ev) {
        String id    = ev.getString("graphId");
        String label = ev.getString("label");
        AbstractGraph old = registry.get(id);
        int newSize = old.getVertexCount() + 1;
        AbstractGraph neo;
        if (old.getRepType() == AbstractGraph.RepType.MATRIX)
            neo = new AdjacencyMatrixGraph(newSize);
        else
            neo = new AdjacencyListGraph(newSize);

        // copia rótulos, pesos e arestas
        for (int i = 0; i < old.getVertexCount(); i++) {
            neo.setVertexLabel(i, old.getVertexLabel(i));
            neo.setVertexWeight(i, old.getVertexWeight(i));
        }
        int newIdx = newSize - 1;
        neo.setVertexLabel(newIdx, label != null ? label : "v" + newIdx);
        // copia arestas
        for (int u = 0; u < old.getVertexCount(); u++)
            for (int v = 0; v < old.getVertexCount(); v++)
                if (old.hasEdge(u, v)) {
                    neo.addEdge(u, v);
                    neo.setEdgeWeight(u, v, old.getEdgeWeight(u, v));
                }
        registry.remove(id);
        registry.register(id, neo);
        ev.setResult(newIdx);
    }

    private void onAddEdge(Event ev) {
        AbstractGraph g = registry.get(ev.getString("graphId"));
        g.addEdge(ev.getInt("u"), ev.getInt("v"));
        ev.setResult(true);
    }

    private void onRemoveEdge(Event ev) {
        AbstractGraph g = registry.get(ev.getString("graphId"));
        g.removeEdge(ev.getInt("u"), ev.getInt("v"));
        ev.setResult(true);
    }

    private void onHasEdge(Event ev) {
        AbstractGraph g = registry.get(ev.getString("graphId"));
        ev.setResult(g.hasEdge(ev.getInt("u"), ev.getInt("v")));
    }

    private void onGetVertexCount(Event ev) {
        ev.setResult(registry.get(ev.getString("graphId")).getVertexCount());
    }

    private void onGetEdgeCount(Event ev) {
        ev.setResult(registry.get(ev.getString("graphId")).getEdgeCount());
    }

    private void onSetVertexWeight(Event ev) {
        AbstractGraph g = registry.get(ev.getString("graphId"));
        g.setVertexWeight(ev.getInt("v"), ev.getDouble("weight"));
        ev.setResult(true);
    }

    private void onGetVertexWeight(Event ev) {
        AbstractGraph g = registry.get(ev.getString("graphId"));
        ev.setResult(g.getVertexWeight(ev.getInt("v")));
    }

    private void onSetEdgeWeight(Event ev) {
        AbstractGraph g = registry.get(ev.getString("graphId"));
        g.setEdgeWeight(ev.getInt("u"), ev.getInt("v"), ev.getDouble("weight"));
        ev.setResult(true);
    }

    private void onGetEdgeWeight(Event ev) {
        AbstractGraph g = registry.get(ev.getString("graphId"));
        ev.setResult(g.getEdgeWeight(ev.getInt("u"), ev.getInt("v")));
    }

    private void onInDegree(Event ev) {
        AbstractGraph g = registry.get(ev.getString("graphId"));
        ev.setResult(g.getVertexInDegree(ev.getInt("v")));
    }

    private void onOutDegree(Event ev) {
        AbstractGraph g = registry.get(ev.getString("graphId"));
        ev.setResult(g.getVertexOutDegree(ev.getInt("v")));
    }

    private void onIsSuccessor(Event ev) {
        AbstractGraph g = registry.get(ev.getString("graphId"));
        ev.setResult(g.isSuccessor(ev.getInt("u"), ev.getInt("v")));
    }

    private void onIsPredecessor(Event ev) {
        AbstractGraph g = registry.get(ev.getString("graphId"));
        ev.setResult(g.isPredecessor(ev.getInt("u"), ev.getInt("v")));
    }

    private void onIsDivergent(Event ev) {
        AbstractGraph g = registry.get(ev.getString("graphId"));
        ev.setResult(g.isDivergent(
            ev.getInt("u1"), ev.getInt("v1"),
            ev.getInt("u2"), ev.getInt("v2")));
    }

    private void onIsConvergent(Event ev) {
        AbstractGraph g = registry.get(ev.getString("graphId"));
        ev.setResult(g.isConvergent(
            ev.getInt("u1"), ev.getInt("v1"),
            ev.getInt("u2"), ev.getInt("v2")));
    }

    private void onIsIncident(Event ev) {
        AbstractGraph g = registry.get(ev.getString("graphId"));
        ev.setResult(g.isIncident(ev.getInt("u"), ev.getInt("v"), ev.getInt("x")));
    }

    private void onIsConnected(Event ev) {
        ev.setResult(registry.get(ev.getString("graphId")).isConnected());
    }

    private void onIsEmpty(Event ev) {
        ev.setResult(registry.get(ev.getString("graphId")).isEmptyGraph());
    }

    private void onIsComplete(Event ev) {
        ev.setResult(registry.get(ev.getString("graphId")).isCompleteGraph());
    }

    private void onExportGephi(Event ev) {
        AbstractGraph g = registry.get(ev.getString("graphId"));
        g.exportToGEPHI(ev.getString("path"));
        ev.setResult(true);
    }
}

File: Projeto_Python_Apendice/grafo/handlers/graph_handler.py
from ..events.event_type import EventType
from ..events.event import Event
from ..events.event_bus import EventBus
from ..core.graph_registry import GraphRegistry
from ..graph.adjacency_list_graph import AdjacencyListGraph
from ..graph.adjacency_matrix_graph import AdjacencyMatrixGraph

class GraphHandler:
    def __init__(self, registry: GraphRegistry):
        self.registry = registry

    def register_all(self, bus: EventBus):
        bus.subscribe(EventType.GRAPH_CREATE, self.on_create)
        bus.subscribe(EventType.GRAPH_ADD_VERTEX, self.on_add_vertex)
        bus.subscribe(EventType.GRAPH_ADD_EDGE, self.on_add_edge)
        bus.subscribe(EventType.GRAPH_REMOVE_EDGE, self.on_remove_edge)
        bus.subscribe(EventType.GRAPH_HAS_EDGE, self.on_has_edge)
        bus.subscribe(EventType.GRAPH_GET_VERTEX_COUNT, self.on_get_vertex_count)
        bus.subscribe(EventType.GRAPH_GET_EDGE_COUNT, self.on_get_edge_count)
        bus.subscribe(EventType.GRAPH_SET_VERTEX_WEIGHT, self.on_set_vertex_weight)
        bus.subscribe(EventType.GRAPH_GET_VERTEX_WEIGHT, self.on_get_vertex_weight)
        bus.subscribe(EventType.GRAPH_SET_EDGE_WEIGHT, self.on_set_edge_weight)
        bus.subscribe(EventType.GRAPH_GET_EDGE_WEIGHT, self.on_get_edge_weight)
        bus.subscribe(EventType.GRAPH_IN_DEGREE, self.on_in_degree)
        bus.subscribe(EventType.GRAPH_OUT_DEGREE, self.on_out_degree)
        bus.subscribe(EventType.GRAPH_IS_SUCCESSOR, self.on_is_successor)
        bus.subscribe(EventType.GRAPH_IS_PREDECESSOR, self.on_is_predecessor)
        bus.subscribe(EventType.GRAPH_IS_DIVERGENT, self.on_is_divergent)
        bus.subscribe(EventType.GRAPH_IS_CONVERGENT, self.on_is_convergent)
        bus.subscribe(EventType.GRAPH_IS_INCIDENT, self.on_is_incident)
        bus.subscribe(EventType.GRAPH_IS_CONNECTED, self.on_is_connected)
        bus.subscribe(EventType.GRAPH_IS_EMPTY, self.on_is_empty)
        bus.subscribe(EventType.GRAPH_IS_COMPLETE, self.on_is_complete)
        bus.subscribe(EventType.GRAPH_EXPORT_GEPHI, self.on_export_gephi)

    def on_create(self, ev: Event):
        gid = ev.get_string("graphId")
        n = ev.get_int("numVertices")
        impl = ev.get_string("impl")
        if self.registry.contains(gid):
            raise ValueError(f"Grafo já existe: {gid}")
        
        if impl == "matrix":
            g = AdjacencyMatrixGraph(n)
        else:
            g = AdjacencyListGraph(n)
        self.registry.register(gid, g)

    def on_add_vertex(self, ev: Event):
        gid = ev.get_string("graphId")
        label = ev.get_string("label")
        old_g = self.registry.get(gid)
        n = old_g.get_vertex_count()
        
        # Cria novo grafo com n+1 vértices
        from ..graph.abstract_graph import RepType
        if old_g.rep_type == RepType.MATRIX:
            new_g = AdjacencyMatrixGraph(n + 1)
        else:
            new_g = AdjacencyListGraph(n + 1)
            
        # Copia dados
        for i in range(n):
            new_g.set_vertex_label(i, old_g.get_vertex_label(i))
            new_g.set_vertex_weight(i, old_g.get_vertex_weight(i))
            for j in range(n):
                if old_g.has_edge(i, j):
                    new_g.set_edge_weight(i, j, old_g.get_edge_weight(i, j))
        
        if label:
            new_g.set_vertex_label(n, label)
        self.registry.register(gid, new_g)

    def on_add_edge(self, ev: Event):
        g = self.registry.get(ev.get_string("graphId"))
        g.add_edge(ev.get_int("u"), ev.get_int("v"))

    def on_remove_edge(self, ev: Event):
        g = self.registry.get(ev.get_string("graphId"))
        g.remove_edge(ev.get_int("u"), ev.get_int("v"))

    def on_has_edge(self, ev: Event):
        g = self.registry.get(ev.get_string("graphId"))
        ev.set_result(g.has_edge(ev.get_int("u"), ev.get_int("v")))

    def on_get_vertex_count(self, ev: Event):
        g = self.registry.get(ev.get_string("graphId"))
        ev.set_result(g.get_vertex_count())

    def on_get_edge_count(self, ev: Event):
        g = self.registry.get(ev.get_string("graphId"))
        ev.set_result(g.get_edge_count())

    def on_set_vertex_weight(self, ev: Event):
        g = self.registry.get(ev.get_string("graphId"))
        g.set_vertex_weight(ev.get_int("v"), ev.get_double("weight"))

    def on_get_vertex_weight(self, ev: Event):
        g = self.registry.get(ev.get_string("graphId"))
        ev.set_result(g.get_vertex_weight(ev.get_int("v")))

    def on_set_edge_weight(self, ev: Event):
        g = self.registry.get(ev.get_string("graphId"))
        g.set_edge_weight(ev.get_int("u"), ev.get_int("v"), ev.get_double("weight"))

    def on_get_edge_weight(self, ev: Event):
        g = self.registry.get(ev.get_string("graphId"))
        ev.set_result(g.get_edge_weight(ev.get_int("u"), ev.get_int("v")))

    def on_in_degree(self, ev: Event):
        g = self.registry.get(ev.get_string("graphId"))
        ev.set_result(g.get_vertex_in_degree(ev.get_int("v")))

    def on_out_degree(self, ev: Event):
        g = self.registry.get(ev.get_string("graphId"))
        ev.set_result(g.get_vertex_out_degree(ev.get_int("v")))

    def on_is_successor(self, ev: Event):
        g = self.registry.get(ev.get_string("graphId"))
        ev.set_result(g.is_successor(ev.get_int("u"), ev.get_int("v")))

    def on_is_predecessor(self, ev: Event):
        g = self.registry.get(ev.get_string("graphId"))
        ev.set_result(g.is_predecessor(ev.get_int("u"), ev.get_int("v")))

    def on_is_divergent(self, ev: Event):
        g = self.registry.get(ev.get_string("graphId"))
        ev.set_result(g.is_divergent(ev.get_int("u1"), ev.get_int("v1"), ev.get_int("u2"), ev.get_int("v2")))

    def on_is_convergent(self, ev: Event):
        g = self.registry.get(ev.get_string("graphId"))
        ev.set_result(g.is_convergent(ev.get_int("u1"), ev.get_int("v1"), ev.get_int("u2"), ev.get_int("v2")))

    def on_is_incident(self, ev: Event):
        g = self.registry.get(ev.get_string("graphId"))
        ev.set_result(g.is_incident(ev.get_int("u"), ev.get_int("v"), ev.get_int("x")))

    def on_is_connected(self, ev: Event):
        g = self.registry.get(ev.get_string("graphId"))
        ev.set_result(g.is_connected())

    def on_is_empty(self, ev: Event):
        g = self.registry.get(ev.get_string("graphId"))
        ev.set_result(g.is_empty_graph())

    def on_is_complete(self, ev: Event):
        g = self.registry.get(ev.get_string("graphId"))
        ev.set_result(g.is_complete_graph())

    def on_export_gephi(self, ev: Event):
        g = self.registry.get(ev.get_string("graphId"))
        g.export_to_gephi(ev.get_string("path"))
